### DISCUSSION TITLE
Testing next.js app - No router instance found

### DISCUSSION THREAD

**Discussion by capdavid** [vercel/next.js#13891] [Help]
Hi, I am trying to test my next.js app using Jest and React testing library. First I got an error 

> TypeError: Cannot read property 'pathname' of null

Which I solved by mocking useRouter like this: 
based on https://github.com/vercel/next.js/issues/7479

```ts
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
export function mockNextUseRouter(pathname: string) {
  useRouter.mockImplementation(() => ({
    route: '',
    basePath: '',
    pathname,
    query: {},
    asPath: '',
    push: async () => true,
    replace: async () => true,
    reload: () => null,
    back: () => null,
    prefetch: async () => undefined,
    beforePopState: () => null,
    isFallback: false,
    events: {
      on: () => null,
      off: () => null,
      emit: () => null,
    },
  }));
}
```

and then inside my test: 

```ts
mockNextUseRouter('/');
it('renders Navbar component', () => {
  render(<Navbar />);
  expect(screen.getByText(/REGISTER/)).toBeInTheDocument();
});
```

But now I want to click on the REGISTER button and check if login modal opens. When I add:
**`fireEvent.click(screen.getByText(/REGISTER/));`**

![image](https://user-images.githubusercontent.com/11329983/84001173-61282280-a966-11ea-8805-4fb71dd73bda.png)

I am very new to testing, can anyone help?

**Comment by proko-sfl** [comment]
Stuck in the exact same situation here. I understand that `Link` uses router internally and router is not available in jest environment but cannot find the way to correctly mock it. We tried a similar mocking of the router but in this case apparently it's not enough. 
Would be great if someone could shed some light on it.

**Comment by capdavid** [comment]
The weird part that I already fixed router not being defined by the first block of code. The second error message is different - I think it has something to do with SSR since it says "You should only use next/router inside the client side..." Maybe Next checks if the app runs on the client, but I don't know how. I tried to mock the Window object, that didn't work. At the least I don't know what property of the window object to mock, (if it'd helped).

**Reply by proko-sfl** [reply]
> I think it has something to do with SSR since it says "You should only use next/router inside the client side..."

My understanding is that the code runs in test environment which is definitely not client side, hence the error message. Even though router itself is mocked, there is still something left to be mocked related to the `<Link>` component. Just don't know what :(

**Comment by sualex** [comment]
@capdavid Thanks for the mock )

**Comment by bistacos** [comment]
This mock worked for me, declared in global scope directly in my components test file:

```javascript
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
useRouter.mockImplementation(() => ({
  pathname: '/',
}));
```

You can add other properties from @capdavid 's mock, in this case `pathname` was the only one I was using. @capdavid Thanks for the mock!

**Reply by RTAndrew** [reply]
This works. I was mocking using the `mockReturnValue`, and changing to `mockImplementation` fixed for me.

**Reply by aroc** [reply]
Fwiw if you're using `import Router from 'next/router';` and calling `Router.push` and other functions outside of components, simply mocking the entire module works fine too if you don't need to make any assertions on router function calls.

`jest.mock("next/router");`

**Reply by deywersonp** [reply]
This 

> This mock worked for me, declared in global scope directly in my components test file:
> 
> ```js
> const useRouter = jest.spyOn(require('next/router'), 'useRouter');
> useRouter.mockImplementation(() => ({
>   pathname: '/',
> }));
> ```
> 
> You can add other properties from @capdavid 's mock, in this case `pathname` was the only one I was using. @capdavid Thanks for the mock!

Worked here! Thx!
